resteasy-jaxrs/src/main/java/org/jboss/resteasy/spi/ResteasyProviderFactory.java

```diff
@@ -258,88 +258,94 @@ public ResteasyProviderFactory(ResteasyProviderFactory parent)
-      contextResolvers = new ConcurrentHashMap<Class<?>, MediaTypeMap<SortedKey<ContextResolver>>>();
-      contextInjectors = new ConcurrentHashMap<Type, ContextInjector>();
-      asyncContextInjectors = new ConcurrentHashMap<Type, ContextInjector>();
-      sortedParamConverterProviders = Collections.synchronizedSortedSet(new TreeSet<ExtSortedKey<ParamConverterProvider>>());
-      stringParameterUnmarshallers = new ConcurrentHashMap<Class<?>, Class<? extends StringParameterUnmarshaller>>();
-      reactiveClasses = new ConcurrentHashMap<Class<?>, Class<? extends RxInvokerProvider<?>>>();
-      headerDelegates = new ConcurrentHashMap<Class<?>, HeaderDelegate>();
+      initialize(null);
+   }
+   
+   protected void initialize(ResteasyProviderFactory parent)
+   {
+      serverDynamicFeatures = parent == null ? new CopyOnWriteArraySet<>() : new CopyOnWriteArraySet<>(parent.serverDynamicFeatures);
+      clientDynamicFeatures = parent == null ? new CopyOnWriteArraySet<>() : new CopyOnWriteArraySet<>(parent.clientDynamicFeatures);
+      enabledFeatures = parent == null ? new CopyOnWriteArraySet<>() : new CopyOnWriteArraySet<>(parent.enabledFeatures);
+      properties = parent == null ? new ConcurrentHashMap<>() : new ConcurrentHashMap<>(parent.properties);
+      providerClasses = parent == null ? new CopyOnWriteArraySet<>() : new CopyOnWriteArraySet<>(parent.providerClasses);
+      providerInstances = parent == null ? new CopyOnWriteArraySet<>() : new CopyOnWriteArraySet<>(parent.providerInstances);
+      classContracts = parent == null ? new ConcurrentHashMap<>() : new ConcurrentHashMap<>(parent.classContracts);
+      serverMessageBodyReaders = parent == null ? new MediaTypeMap<>() : parent.serverMessageBodyReaders.clone();
+      serverMessageBodyWriters = parent == null ? new MediaTypeMap<>() : parent.serverMessageBodyWriters.clone();
+      clientMessageBodyReaders = parent == null ? new MediaTypeMap<>() : parent.clientMessageBodyReaders.clone();
+      clientMessageBodyWriters = parent == null ? new MediaTypeMap<>() : parent.clientMessageBodyWriters.clone();
+      sortedExceptionMappers = parent == null ? new ConcurrentHashMap<>() : new ConcurrentHashMap<>(parent.sortedExceptionMappers);
+      asyncResponseProviders = parent == null ? new ConcurrentHashMap<>() : new ConcurrentHashMap<>(parent.asyncResponseProviders);
+      asyncClientResponseProviders = parent == null ? new ConcurrentHashMap<>() : new ConcurrentHashMap<>(parent.asyncClientResponseProviders);
+      asyncStreamProviders = parent == null ? new ConcurrentHashMap<>() : new ConcurrentHashMap<>(parent.asyncStreamProviders);
+      contextResolvers = new ConcurrentHashMap<>();
+      if (parent != null)
+      {
+         for (Map.Entry<Class<?>, MediaTypeMap<SortedKey<ContextResolver>>> entry : parent.getContextResolvers()
+               .entrySet())
+         {
+            contextResolvers.put(entry.getKey(), entry.getValue().clone());
+         }
+      }

```

Comment:
@ronsigal, because we need the map values to be cloned too and ConcurrentHashMap.clone() does not clone them. Actually I just copied existing code from the previous class version https://github.com/resteasy/Resteasy/blob/f5947196796d0ca0ace15e440d3ce4f4a838349f/resteasy-jaxrs/src/main/java/org/jboss/resteasy/spi/ResteasyProviderFactory.java#L1152

Reply:
@NicoNes, Hmmm, this is interesting. I see now that AbstractMap.clone() "Returns a shallow copy".  But maybe that's OK. Suppose you add a MyContextResolver to a parent config and then create a child config by shallow cloning. There are two separate maps, so removing your MyContextResolver from the parent won't affect the child and vice versa. But what if you do something to reconfigure your MyContextResolver? The spec says

>Therefore, once a configuration is inherited it is also detached (deep copied)
>from its parent configuration and changes to the parent and child configurations are not be visible to each other.

Does that mean reconfiguring the MyContextResolver in the parent shouldn't affect the MyContextResolver in the child? It's a little unclear, but I guess I would lean toward cloning the MyContextResolver. Well, it does say "deep copied".

But then shouldn't all of the ConcurrentHashMaps be deep copied?

Reply:
@ronsigal , many points to discuss here:

> I see now that AbstractMap.clone() "Returns a shallow copy". But maybe that's OK

For our specific case at line https://github.com/resteasy/Resteasy/blob/38b18556acad9d19a9f3d6bee830a8a3002d27f4/resteasy-jaxrs/src/main/java/org/jboss/resteasy/spi/ResteasyProviderFactory.java#L308, we have to perform this loop instead of invoking ConcurrentHashMap.clone() in order to clone all  MediaTypeMap entries so that MediaTypeMap entries  from child configuration and the ones from parent configuration are separated.
So in this case ConcurrentHashMap.clone() is not OK to ensure that modification to child won't be visible to parent and vice versa, since their MediaTypeMap entries will be shared.

> Does that mean reconfiguring the MyContextResolver in the parent shouldn't affect the MyContextResolver in the child?

To my mind it does not mean that.
In my opinion it just means that registration/deregistration on detached configurations must not be visible to each other. But it does not mean that the provider stored in each configuration should not be shared.

To my mind it will be undoable to perform a deep copy of a user provider such as a Filter or a MessageBodyReader for example. And by the way those providers do not extends Cloneable nor upgrade clone() method visibility to public, so we can't clone() them.
If you take a look at MediaTypeMap.clone() and SubtypeMap.clone(),  they does not perform a deep copy of ContextResolver<> instances neither. They just add them to detached collections.

>  But then shouldn't all of the ConcurrentHashMaps be deep copied?

IMO, unless we have a clear and easy way to perfom Provider instance deep copy, making a deep copy of  ConcurrentHashMap is not necessary and actually not possible.

Reply:
@NicoNes: "making a deep copy of ConcurrentHashMap is not necessary and actually not possible."

Good point. Ok.